README.md
Bitterness
==========
[![Build Status](https://travis-ci.org/codepr/bitterness.svg?branch=master)](https://travis-ci.org/codepr/bitterness)

A basic library providing utility methods to calculate bitterness of a given recipe, in IBU measure unit, featuring Rager forumula, Tinseth formula or an average of the two, Garetz formula is also available. Measure system is defaulted at European decimal system, supports Imperial system too.

### Installation
```sh
npm install bitterness
# for development purpose install (e.g. running tests) need mocha and chai
npm install bitterness --dev
```
### Usage
```js
var bt = require('bitterness');
// ibu value with Rager formula
var rager = bt.rager(
		hopWeight, // grams of hop
		time, // time in minutes
		alphaAcids, // AA% in the form of % (e.g. 6% => 6)
		batchSize, // size of the batch in liters
		originalGravity); // OG in the form xxxx.xx (e.g. 1050.00)
// ibu value with Tinseth formula
var tinseth = bt.tinseth(
		hopWeight, // grams of hop
		time, // time in minutes
		alphaAcids, // AA% in the form of % (e.g. 6% => 6)
		batchSize, // size of the batch in liters
		originalGravity); // OG in the form xxxx.xx (e.g. 1050.00)
// ibu value using Garetz formula
var garetz = bt.garetz(
		hopWeight, // grams of hop
		time, // time in minutes
		alphaAcids, // AA% in the form of % (e.g. 6% => 6)
		batchSize, // size of the batch in liters
		originalGravity, // OG in the form xxxx.xx (e.g. 1050.00)
		final_volume // size of the volume post boil in liters
		height); // height in meters of the boiling batch

```
### Practical example:
```js
var bt = require('bitterness');
var rager = bt.rager;
// 93 grams, 90 minutes boil time, 6% alpha acids, 25 liters batch, 1050 og
var r = rager(93, 90, 6, 25, 1050);
var g = bt.garetz(93, 90, 6, 25, 1050); // using defaulted final volume = (batch size - 10%) and height = 0

console.log(r); // ~ 71 ibu
console.log(g); // ~ 49 ibu

bt.setMetricSystem(false); // set system to imperial
// 3.17 oz, 90 minutes, 6% alpha acids, 6.60 gallons, 1050 og
var r = rager(3.17, 90, 6, 6.60, 1050);

```

### Test
```sh
npm test
```
### Changelog

See the [CHANGELOG](CHANGELOG.md) file.

### License

See the [LICENSE-MIT](LICENSE-MIT) file for license rights and limitations (MIT).


CHANGELOG.md
#### 1.0.4 - 2015-11-15
###### Added
- Garetz utilization table.

###### Changed
- Refactored code, better comments.
- Strict mode.

###### Removed
- Ibus additions array
- Average method.
- Cache system.

#### 1.0.2 -  2015-03-20
###### Changed
- Corrected some bugs.

#### 1.0.0 -  2015-03-20
###### Added
- Metric system switch (European/Imperial).

#### 0.1.9 -  2015-03-20
###### Added
- Cache system.
- Last addition.
- Subtotal ibu addition.

###### Changed
- Corrected Tinseth formula.

#### 0.1.7 -  2015-03-19
###### Added
- Garetz method for IBU calculation.
- Garetz method inside average method with optional arguments.

#### 0.1.6 -  2015-03-19
###### Changed
- Corrected some bugs.

#### 0.1.5 -  2015-03-19
###### Changed
- Refactored code into module pattern design.
- Update usage section.

#### 0.1.2 -  2015-03-18
###### Added
- Tests.

###### Changed
- Privatized some utility functions.

#### 0.1.0 -  2014-12-05
###### Added
- Initial release.


test/bitterness.js
var should = require('chai').should(),
    bitterness = require('../index.js'),
    rager = bitterness.rager,
    tinseth = bitterness.tinseth;

var hop_grams = 93,
    alpha_acids = 6,
    time = 90,
    original_gravity = 1050,
    batch_size = 25;

bitterness_1 = new bitterness(hop_grams, time, alpha_acids, batch_size, original_gravity);
var bitterness_2 = new bitterness(70, 80, 8.5, 35, 1057);

describe('#rager', function() {
    it('rager ibu calculation: 71', function() {
        bitterness_1.rager().should.equal(71);
    });

    it('rager ibu calculation: 54', function() {
        bitterness_2.rager().should.equal(54);
    });
});

describe('#tinseth', function() {
    it('tinseth ibu calculation: 135', function() {
        bitterness_1.tinseth().should.equal(135);
    });

    it('tinseth ibu calculation: 108', function() {
        bitterness_2.tinseth().should.equal(108);
    });

});

describe('#average', function() {
    it('average ibu calculation: 103', function() {
        bitterness_1.average().should.equal(103);
    });

    it('average ibu calculation: 81', function() {
        bitterness_2.average().should.equal(81);
    });
});


test/bitterness.test.js
var should = require('chai').should();
var bt = require('../index.js');

describe('Bitterness calculations according to European metric system:\n', function() {

	describe('# w: 93 grams, t: 90 mins, aa: 6%, b: 25 lit, og: 1050', function() {

		it('Rager: 71.26', function() {
			bt.rager(93, 90, 6, 25, 1050).should.equal(71.26);
		});

		it('Tinseth: 61.19', function() {
			bt.tinseth(93, 90, 6, 25, 1050).should.equal(61.19);
		});

		it('Garetz: 48.59', function() {
			bt.garetz(93, 90, 6, 25, 1050, 23, 0).should.equal(48.59);
		});

	});

	describe('# w: 70 grams, t: 80 mins, aa: 8.5%, b: 35 lit, og: 1057', function() {

		it('Rager: 53.93', function() {
			bt.rager(70, 80, 8.5, 35, 1057).should.equal(53.93);
		});

		it('Tinseth: 43.16', function() {
			bt.tinseth(70, 80, 8.5, 35, 1057).should.equal(43.16);
		});

		it('Garetz: 35.94', function() {
			bt.garetz(70, 80, 8.5, 35, 1057, 32, 0).should.equal(35.94);
		});
	});
});

describe('Bitterness calculations according to Imperial metric system:\n', function() {
	describe('# w: 3.17 oz, t: 90 mins, aa: 6%, b: 6.6 gal, og: 1050', function() {

		it('Rager: 68.91', function() {
			bt.setMetricSystem(false);
			bt.rager(3.17, 90, 6, 6.60, 1050).should.equal(68.91);
		});

		it('Tinseth: 59.18', function() {
			bt.tinseth(3.17, 90, 6, 6.60, 1050).should.equal(59.18);
		});

		it('Garetz: 48.45', function() {
			bt.garetz(3.17, 90, 6, 6.60, 1050).should.equal(48.45);
		});
	});

	describe('# w: 3.17 oz, t: 90 mins, aa: 6%, b: 6.6 gal, og: 1050, fb: 5.94, height: 6 feets', function() {

		it('Rager: 68.91', function() {
			bt.rager(3.17, 90, 6, 6.60, 1050).should.equal(68.91);
		});

		it('Tinseth: 59.18', function() {
			bt.tinseth(3.17, 90, 6, 6.60, 1050).should.equal(59.18);
		});

		it('Garetz: 48.42', function() {
			bt.garetz(3.17, 90, 6, 6.60, 1050, 5.94, 6).should.equal(48.42);
		});
	});
});


lib/bitterness.js
/**
 * Bitterness
 * https://github.com/codepr/bitterness
 *
 * Copyright (c) 2014 Andrea Baldan
 * Licensed under the MIT license.
 **/

'use strict'

var bitterness = module.exports = (function() {

	var DECIMAL_METRIC_FACTOR = 7490;
	var IMPERIAL_METRIC_FACTOR = 74.90;
	// constants derived from respective methods
	// Rager constants
	var STANDARD_BASE_OG = 1050;
	var RAGER_BOIL_TIME_ADJUSTMENT = 31.32;
	// Tinseth constants
	// accordingly to Glenn Tinseth, the numbers 0.000125 and 1.65 are
	// empirically derived to fit his data.
	// The number 0.04 controls the shape of the utils vs time curve, finally
	// 4.15 controls the max util value, should be adjusted based on personal
	// util value.
	var TINSETH_BIGNESS_GRAVITY_BASE = 0.000125;
	var TINSETH_BIGNESS_MULTIPLIER = 1.65;
	var TINSETH_CURVE_SHAPE = -0.04;
	var TINSETH_MAX_UTILIZATION = 4.15;
	// Garetz constants
	var GARETZ_HEIGHT_METERS = 168;
	var GARETZ_HEIGHT_FEETS = 550;

	var isMetric = true;

	//Tanh functions for full compatibility, apparently node before 0.11 doesn't
	//support Math.tanh
	//
	var tanh = function(arg) {
		return (Math.exp(arg) - Math.exp(-arg)) / (Math.exp(arg) + Math.exp(-arg));
	}

	// alpha acids utilization vs time, according to Rager can be reduced to this
	// equation

	var util = function(time) {
		return 18.11 + 13.86 * tanh((time - RAGER_BOIL_TIME_ADJUSTMENT) / 18.27);
	}

	// Garetz formula for utilization take in account yeast flocculation, these
	// ranges represent average flocculation

	var garetzUtil = function(time) {
		if (time < 11) {
			return 0;
		} else if (time > 10 && time < 16) {
			return 2;
		} else if (time > 15 && time < 21) {
			return 5;
		} else if (time > 20 && time < 26) {
			return 8;
		} else if (time > 25 && time < 31) {
			return 11;
		} else if (time > 30 && time < 36) {
			return 14;
		} else if (time > 35 && time < 41) {
			return 16;
		} else if (time > 40 && time < 46) {
			return 18;
		} else if (time > 45 && time < 51) {
			return 19;
		} else if (time > 50 && time < 61) {
			return 20;
		} else if (time > 60 && time < 71) {
			return 21;
		} else if (time > 70 && time < 81) {
			return 22;
		} else {
			return 23;
		}
	}

	// according to Rager, if the gravity of the boil exceeds 1050 there is a
	// gravity adjustment to factor in

	var adjustment = function(originalGravity) {
		if(originalGravity > STANDARD_BASE_OG) {
			return ((originalGravity - STANDARD_BASE_OG) / 2) * 0.001;
		}
		else return 0;
	}

	// mg/l of added alpha acids needed in the formula IBUs = decimal util * mg/l

	var mgAlphaAcids = function(hopGrams, alphaAcids, batchSize, metric) {
		var factor = 1000;
		if(metric == false) factor = DECIMAL_METRIC_FACTOR;
		return ((hopGrams * (alphaAcids * 0.01) * factor) / (batchSize - (batchSize * 0.1)));
	}

	// empirically retrieved factors used to calculate decimal util
	// bigness factor..

	var bignessFactor = function(originalGravity) {
		return (TINSETH_BIGNESS_MULTIPLIER *
				Math.pow(TINSETH_BIGNESS_GRAVITY_BASE, ((originalGravity * 0.001) - 1)));
	}

	// ..and boiling time factor

	var boilingTimeFactor = function(time) {
		return (1 - Math.exp(TINSETH_CURVE_SHAPE * time)) / TINSETH_MAX_UTILIZATION;
	}

	// Garetz method expect some factors to be combined together
	// based on height and temperature..

	var temperatureFactor = function(height) {
		var heightFactor = GARETZ_HEIGHT_METERS;
		if(isMetric == false) heightFactor = GARETZ_HEIGHT_FEETS;
		return ((height / heightFactor) * 0.02) + 1;
	}

	// .. and gravity based on different state of volume

	var gravityFactor = function(finalVolume, currentVolume, originalGravity) {
		var boilGravity = ((finalVolume / currentVolume) * ((originalGravity * 0.001) - 1)) + 1;
		return (5 * (boilGravity - 0.85));
	}

	return {

		/**
		 * Calculate bitterness of a given recipe in IBU using Rager formula.
		 *
		 * @param {number} hopWeight hop weight in grams / oz according to metric
		 * system
		 * @param {number} time time in minutes
		 * @param {number} alphaAcids alpha acids in % form (e.g. 6% => 6)
		 * @param {number} batchSize batch size in liters / gal according to metric
		 * system
		 * @param {number} originalGravity original gravity (xxxx.xx format e.g.
		 * 1050.00)
		 * @return {number} Number of IBU according to Rager formula
		 **/

		rager: function(hopWeight, time, alphaAcids, batchSize, originalGravity) {
			var _this = this;
			var result = 0;
			var factor = 10;
			if(_this.isMetric == false) factor = IMPERIAL_METRIC_FACTOR;
			result = (hopWeight * util(time) * alphaAcids * factor) /
				(batchSize * (1 + adjustment(originalGravity)));
			result = result / 100;
			return Math.round(result * 100) / 100;
		},

		/**
		 * Calculate bitterness of a given recipe in IBU using Tinseth formula.
		 *
		 * @param {number} hopWeight hop weight in grams / oz according to metric
		 * system
		 * @param {number} time time in minutes
		 * @param {number} alphaAcids alpha acids in % form (e.g. 6% => 6)
		 * @param {number} batchSize batch size in liters / gal according to metric
		 * system
		 * @param {number} originalGravity original gravity (xxxx.xx format e.g.
		 * 1050.00)
		 * @return {number} Number of IBU according to Tinseth formula
		 **/

		tinseth: function(hopWeight, time, alphaAcids, batchSize, originalGravity) {
			var _this = this;
			var result = 0;
			var mg_alpha = mgAlphaAcids(hopWeight, alphaAcids, batchSize, _this.isMetric);
			var b_factor = bignessFactor(originalGravity);
			var bt_factor = boilingTimeFactor(time);
			result = b_factor * bt_factor * mg_alpha;
			return Math.round(result * 100) / 100;
		},

		/**
		 * Calculate bitterness of a given recipe in IBU using Garetz formula.
		 *
		 * @param {number} hopWeight hop weight in grams / oz according to metric
		 * system
		 * @param {number} time time in minutes
		 * @param {number} alphaAcids alpha acids in % form (e.g. 6% => 6)
		 * @param {number} batchSize batch size in liters / gal according to metric
		 * system
		 * @param {number} originalGravity original gravity (xxxx.xx format e.g.
		 * 1050.00)
		 * @param {number} finalVolume final volume in liters / gal according to
		 * metric system
		 * @param {number} height height of the boiling batch in meters / feets
		 * according to metric system
		 * @return {number} Number of IBU according to Garetz formula
		 **/

		garetz: function(hopWeight, time, alphaAcids, batchSize, originalGravity, finalVolume, height) {
			var _this = this;
			var result = 0;
			var factor = 1;
			var _finalVolume = finalVolume || (batchSize - (batchSize * 0.1));
			var _height = height || 0;
			var gf = gravityFactor(_finalVolume, batchSize, originalGravity);
			var tf = temperatureFactor(_height);
			if(_this.isMetric == false) factor = 7.490;
			var fr = (garetzUtil(time) * alphaAcids * hopWeight * factor) /
				(650 * batchSize * gf * tf);
			result = (130 * (-1 + (Math.sqrt(1 + fr))) / (_finalVolume / batchSize));
			return Math.round(result * 100) / 100;
		},

		/**
		 * Set preferred measure unit system (default is true, e.g. European)
		 * @param {boolean} metric true or false
		 */

			setMetricSystem: function(metric) {
				if(typeof(metric)== 'boolean') this.isMetric = metric;
			}
	};
})();
